Add user's assigned locations to invitation object and import/export

Robert-Costello opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
The user facing portion added here is the ability to import/export web users with all assigned locations in addition to primary location.
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
This is a model change for Invitation that is part of the work to allow assigning locations as part of the web user invite work flow.
[Jira ticket](https://dimagi.atlassian.net/browse/USH-4507)
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally and on staging. Has also been through QA successfully.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
[QA Ticket](https://dimagi.atlassian.net/browse/QA-6367)

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
The code depends on the renamed `primary_location` field. 

Making the migration independent of the code will require multiple PRs, a management command to copy values between fields, and changelog for third party hosts. This is rather complicated and a rename in a single PR is faster and simpler than copying data between two columns. The rename will lock the table, so no data will get read/written from the wrong place, but users might encounter transient errors during deploy. Invitations is not a primary workflow so I am inclined to accept the risk of transient errors during deploy.

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
This PR renames model fields so a reversion of this PR would require also reverting any changes that depend on the change field names.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Robert-Costello committed:
model change and migration

Robert-Costello committed:
update method and uses of add_as_web_user

Robert-Costello committed:
update bulk download and test

Jtang-1 commented on corehq/apps/users/models.py:
```diff
@@ -531,14 +531,17 @@ def add_domain_membership(self, domain, timezone=None, **kwargs):
 
-    def add_as_web_user(self, domain, role, location_id=None, program_id=None):
+    def add_as_web_user(self, domain, role, primary_location_id=None,
+                        assigned_locations_ids=[], program_id=None):
```
I think it's best practice to not have mutable objects as default values https://docs.python.org/3/reference/compound_stmts.html#function-definitions

> This is especially important to understand when a default parameter value is a mutable object, such as a list or a dictionary: if the function modifies the object (e.g. by appending an item to a list), the default parameter value is in effect modified. This is generally not what was intended. A way around this is to use None as the default, and explicitly test for it in the body of the function, e.g.:

So should assign default value as none and in the body do 
```
if assigned_locations_ids is None:
    assigned_locations_id = []
```

Jtang-1 committed:
rename varibles for Invitation field rename + add_as_web_user parameter name change

Jtang-1 committed:
align variable naming with rest of the codebase

I do think assigned_locations_ids is more correct. But for the sake
of consistency

Jtang-1 committed:
bug: SQLLocation field is location_id

Jtang-1 committed:
bug: Many to Many fields need to be accessed via .all()

Robert-Costello commented on corehq/apps/users/bulk_download.py:
```diff
@@ -203,7 +203,8 @@ def make_web_user_dict(user, location_cache, domain):
     try:
-        location_codes.append(location_cache.get(getattr(invite.location, 'location_id', None)))
+        location_codes.append(location_cache.get(getattr(invite.primary_location, 'location_id', None)))
+        location_codes.append(location_cache.get(getattr(invite.assigned_locations, 'location_id', None)))
```
sorry, I was assuming we'd incorporate a loop. could have been more explicit there. 
good point about the queryset, I'll make a change with that in mind.

Robert-Costello committed:
use default of None

Robert-Costello committed:
use all() and extract location ids from queryset

Jtang-1 committed:
Merge branch 'master' into rc/assigned-locations-invite-migration

Jtang-1 committed:
rename parameter name for clarity

Jtang-1 committed:
feat: supports assigned_location_ids for web user imports

Jtang-1 committed:
uses assigned_locations upon creation of webuser from import

Jtang-1 committed:
bug: variable needs initial value

Jtang-1 committed:
test: update to test assigned locations are included

Jtang-1 committed:
make primary and assigned locations optional parameters

Jtang-1 commented on corehq/apps/user_importer/importer.py:
```diff
@@ -319,7 +319,7 @@ def get_location_from_site_code(site_code, location_cache):
 
 def create_or_update_web_user_invite(email, domain, role_qualified_id, upload_user, primary_location_id,
-                                     user_change_logger=None, send_email=True):
+                                    assigned_location_ids, user_change_logger=None, send_email=True):
```
[0c86eb2](https://github.com/dimagi/commcare-hq/pull/34550/commits/0c86eb249798f4c035c21eccda4fb5e3e2bae560)

Jtang-1 commented:
Getting final confirmation from QA on some non-issue tickets but if that ends up being true then this has passed QA. Bumping for review

AddisonDunn reviewed (commented):
Only one comment of potential substance.

Jtang-1 committed:
bug: removes repeat of the primary location since it is also a member of assigned_location

Jtang-1 commented on corehq/apps/users/bulk_download.py:
```diff
@@ -203,7 +203,10 @@ def make_web_user_dict(user, location_cache, domain):
     location_codes = []
     try:
-        location_codes.append(location_cache.get(getattr(invite.location, 'location_id', None)))
+        location_codes.append(location_cache.get(getattr(invite.primary_location, 'location_id', None)))
```
The only change here was the property rename `location` to `primary_location` so the existing behavior adds `None` to the list so I'm not expecting there to be an issue. I also tested it and didn't see an issue.

But i did catch a different bug while investigating this [1e10e35](https://github.com/dimagi/commcare-hq/pull/34550/commits/1e10e35029388f7a17855387f33776427ae433f6) so that was lucky

Jtang-1 replied:
I think it existed longer, i was looking at this https://github.com/dimagi/commcare-hq/blame/597a4f3f460c14368db47ff2649dfee0d56d63ef/corehq/apps/users/bulk_download.py#L203-L206 

Steph added this line`location_codes.append(location_cache.get(invite.supply_point))` 3 years ago, and `invite.supply_point` could be `None`

Robert-Costello commented on corehq/apps/user_importer/importer.py:
```diff
@@ -318,19 +318,25 @@ def get_location_from_site_code(site_code, location_cache):
     )
+    assigned_locations = [SQLLocation.by_location_id(assigned_location_id)
+            for assigned_location_id in assigned_location_ids
+            if assigned_location_id is not None]
```
right, so if the `assigned_location_ids` is set to an empty list, this loop won't iterate over anything. that's what I think makes this check redundant. I don't think there is any risk of `assigned_location_ids` containing `None` values.

Jtang-1 replied:
I think the list actually could contain `None` for example [here](https://github.com/dimagi/commcare-hq/blob/0c86eb249798f4c035c21eccda4fb5e3e2bae560/corehq/apps/user_importer/importer.py#L787-L796) `user_invite_locs_ids` would contain none is there's no location specified for that row (I'm pretty sure, but something's wrong with my computer when i try to upload a file so haven't been able to test it)

Robert-Costello replied:
I could be wrong but it seems like the check for len of `location_codes` here https://github.com/dimagi/commcare-hq/blob/0c86eb249798f4c035c21eccda4fb5e3e2bae560/corehq/apps/user_importer/importer.py#L785
and the fact that `get_location_from_site_code` will throw an exception if something other than an int or string is passed in means that there wouldn't be `None` values in `user_invite_locs_ids`.
https://github.com/dimagi/commcare-hq/blob/885be244e5071386e1f73e4229d80421d1ed0cbf/corehq/apps/user_importer/importer.py#L301-L310
I definitely don't consider this blocking, fyi.

Jtang-1 replied:
Ah i see, agreed that makes sense. [5c60667](https://github.com/dimagi/commcare-hq/pull/34550/commits/5c6066798694b39189b17df8c5f5df65f7d87d4c)

Jtang-1 committed:
simplify by using values list instead of list comprehension

Jtang-1 commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -308,13 +308,18 @@ def _fmt_phone(phone_number):
             self._log_primary_location_info()
+        if assigned_location_ids:
+            assigned_locations = self.user.get_sql_locations(self.user_domain)
+            self.logger.add_info(UserChangeMessage.assigned_locations_info(assigned_locations))
```
I considered making a function like that, but i couldn't figure out a way to combine the logic with the logging [here](https://github.com/dimagi/commcare-hq/blob/a2d52986c9e85d0baee4efe4fc57c20b176da192/corehq/apps/user_importer/helpers.py#L354-L361). So I thought it would be confusing the introduce a function that looks like it should but couldn't be used in both spots

Jtang-1 commented on corehq/apps/users/models.py:
```diff
@@ -2819,7 +2826,8 @@ def accept_invitation_and_join_domain(self, web_user):
             role=self.role,
-            location_id=getattr(self.location, "location_id", None),
+            primary_location_id=getattr(self.primary_location, "location_id", None),
+            assigned_location_ids=[getattr(loc, "location_id", None) for loc in self.assigned_locations.all()],
```
good clean up - [8baf3ce](https://github.com/dimagi/commcare-hq/pull/34550/commits/8baf3cedfb9ee0efcb9566a0c26c97e230fd6bf0) Tested it locally too

Jtang-1 commented on corehq/apps/users/management/commands/accept_invite.py:
```diff
@@ -25,7 +25,9 @@ def handle(self, username, domain, **options):
         user.add_as_web_user(invitation.domain, role=invitation.role,
-                            location_id=getattr(invitation.location, "location_id", None),
+                            primary_location_id=getattr(invitation.primary_location, "location_id", None),
+                            assigned_location_ids=[getattr(loc, "location_id", None)
```
Same as below, https://github.com/dimagi/commcare-hq/pull/34550#discussion_r1594386807

Jtang-1 committed:
assigned_location_ids is not expected to contain None as a value
